cleverhans/model_abs.py

```diff
@@ -0,0 +1,138 @@
+from abc import ABCMeta
+
+
+class Model(object):
+    """
+    An abstract interface for model wrappers that exposes model symbols
+    needed for making an attack. This abstraction removes the dependency on
+    any specific neural network package (e.g. Keras) from the core
+    code of CleverHans. It can also simplify exposing the hidden features of a
+    model when a specific package does not directly expose them (needed by
+    "features adversaries").
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, model):
+        """
+        Init a wrapper. If `fprop` is implemented, `__init__`
+        should keep track of the name of the layers or `self.model` should
+        provide a method for retrieving a layer.
+
+        :param model: A function that takes a symbolic input and returns the
+                      symbolic output for the model's predictions.
+        """
+
+        pass
+
+    def fprop(self, x, layer=None):
+        """
+        Expose the hidden features of a model given a layer name.
+
+        :param x: A symbolic representation of the network input
+        :param layer: The name of the hidden layer to return features at.
+        :return: A symbolic representation of the hidden features
+        """
+        error = 'Feature extraction for hidden layers not implemented'
+        raise NotImplementedError(error)
+
+    def get_logits(self, x):
+        """
+        :param x: A symbolic representation of the network input
+        :return: A symbolic representation of the output logits, values before
+                 softmax.
+        """
+        error = '`get_logits` not implemented'
+        raise NotImplementedError(error)
+
+    def get_probs(self, x):
+        """
+        :param x: A symbolic representation of the network input
+        :return: A symbolic representation of the output probabilities, values
+                 after softmax.
+        """
+        error = '`get_probs` not implemented'
+        raise NotImplementedError(error)
+
+
+class KerasModelWrapper(Model):
+    """
+    An implementation of ModelAbstraction that wraps a Keras model. It
+    specifically exposes the hidden features of a model by creating new models.
+    The symbolic graph is reused and so there is little overhead. Splitting
+    in-place operations can incur an overhead.
+    """
+
+    def __init__(self, model):
+        """
+        Create a wrapper for a Keras model
+
+        :param model: A Keras model
+        """
+        super(KerasModelWrapper, self).__init__(model)
+
+        # Initialize attributes
+        self.model = model
+        self.model_dict = {None: model}
```

Comment:
Not all wrappers need to have caching. Keras models need it because they don't expose the hidden features themselves and we have to create a new model to expose a specific layer. But other models such as pure tf models don't necessarily need it because no new object is created.